encryption/encryption.cc

```diff
@@ -0,0 +1,466 @@
+                     reinterpret_cast<const unsigned char*>(key.data()), iv,
+                     static_cast<int>(encrypt_type)),
+      "Failed to init cipher.");
+
+  // Disable padding. After disabling padding, data size should always be
+  // multiply of kBlockSize.
+  OPENSSL_RET_NOT_OK(EVP_CIPHER_CTX_set_padding(ctx.get(), 0),
+                     "Failed to disable padding for cipher context.");
+
+  uint64_t data_offset = 0;
+  size_t remaining_data_size = data_size;
+  int output_size = 0;
+  unsigned char partial_block[kBlockSize];
+
+  // In the following we assume EVP_CipherUpdate allow in and out buffer are
+  // the same, to save one memcpy. This is not specified in official man page.
+
+  // Handle partial block at the beginning. The partial block is copied to
+  // buffer to fake a full block.
+  if (block_offset > 0) {
+    size_t partial_block_size =
+        std::min<size_t>(kBlockSize - block_offset, remaining_data_size);
+    memcpy(partial_block + block_offset, data, partial_block_size);
+    OPENSSL_RET_NOT_OK(
+        EVP_CipherUpdate(ctx.get(), partial_block, &output_size, partial_block,
+                         static_cast<int>(kBlockSize)),
+        "Crypter failed for first block, offset " +
+            std::to_string(file_offset));
+    if (UNLIKELY(output_size != static_cast<int>(kBlockSize))) {
+      return Status::IOError(
+          "Unexpected crypter output size for first block, expected " +
+          std::to_string(kBlockSize) + " vs actual " +
+          std::to_string(output_size));
+    }
+    memcpy(data, partial_block + block_offset, partial_block_size);
+    data_offset += partial_block_size;
+    remaining_data_size -= partial_block_size;
+  }
+
+  // Handle full blocks in the middle.
+  if (remaining_data_size >= kBlockSize) {
+    size_t actual_data_size =
+        remaining_data_size - remaining_data_size % kBlockSize;
+    unsigned char* full_blocks =
+        reinterpret_cast<unsigned char*>(data) + data_offset;
+    OPENSSL_RET_NOT_OK(
+        EVP_CipherUpdate(ctx.get(), full_blocks, &output_size, full_blocks,
+                         static_cast<int>(actual_data_size)),
+        "Crypter failed at offset " +
+            std::to_string(file_offset + data_offset));
+    if (UNLIKELY(output_size != static_cast<int>(actual_data_size))) {
+      return Status::IOError("Unexpected crypter output size, expected " +
+                             std::to_string(actual_data_size) + " vs actual " +
+                             std::to_string(output_size));
+    }
+    data_offset += actual_data_size;
+    remaining_data_size -= actual_data_size;
+  }
+
+  // TODO(yingchun): Can we remove the end partial block handling if handling a
+  //  suitable adjusted partial block at the beginning and full blocks in the
+  //  middle?
+  // Handle partial block at the end. The partial block is copied to buffer to
+  // fake a full block.
+  if (remaining_data_size > 0) {
```

Comment:
Can use assert to check whether `remaining_data_size` is 0.

Reply:
No, it's a valid case while remaining_data_size > 0.

Reply:
yes, but I mean use `assert` to replace `if`.

Reply:
Both cases (remaining_data_size > 0 and remaining_data_size == 0) are valid.